dosbox-staging/dosbox-staging
# Improve language (translation) autodetection

# Description

- some minor refactoring/cleanup
- improved language autodetection, see the _Release notes_ proposal

# Release notes

Language (translation) autodetection now checks the host OS list of user preferred languages (not just the current GUI language) - if it can't find the translation into the most preferred language, it tries with the second one from the list, and so on.

# Manual testing

1. Start DOSBox by command `dosbox -lang en`, `dosbox -lang pl`, etc. - check that the correct translation is loaded (pay attention to the first `LOCALE` log). If the startup banner is hidden, but the translation is up to date, a localized text is probably (depending on the configuration) visible even on the DOSBox window title bar.
If unknown language is specified, it should use internal English; `dosbox -lang auto` should trigger language autodetection (see the other steps).
The command line option should have a priority over `language = ` setting from the config file.

2. Set `language = en`, `language = pl`, etc. - check that the correct translation is loaded when DOSBox is started. If unknown language is specified, it should use internal English.

3. **[Linux]**  For this test set `language = auto` in the config file.
Modern Linux GUIs do not allow to set language priorities, we have to manually set environment variables for testing - see https://www.gnu.org/software/gettext/manual/html_node/The-LANGUAGE-variable.html for the `LANGUAGE` variable description (it allows to specify multiple languages and has a priority over `LC_ALL`, `LC_MESSAGES` and `LANG` variables).
Example commands:
- `LANGUAGE=pl LC_ALL=de dosbox` - should select the Polish translation.
- `LANGUAGE=pl_PL:cz_CZ LC_ALL=de dosbox` - should select the Polish translation.
- `LANGUAGE=cz_CZ:pl_US LC_ALL=de dosbox` - should select the Polish translation (we have no Czech messages).
- `LANGUAGE= LC_ALL=de dosbox` - should select the German translation.
- `LANGUAGE=cz LC_ALL=de dosbox` - should select  the internal English translation (on POSIX systems the higher priority variables override the lower priority ones).

4.  **[macOS]**  For this test set `language = auto` in the config file.
- Press the Apple logo, _System Settings..._ -> _General_ -> _Language & Region_.
- Manipulate the _Preferred Languages_ list - add/remove/reorder (you don't have to reboot).
BE REASONABLE. If you end up with your Macintosh communicating in Sindarin and now you have no idea how to get back - this is not my fault :D
- Start DOSBox - it should use the first language from the list we have a translation for.

5.  **[Windows 11]**  For this test set `language = auto` in the config file.
- Start the _Settings_ app, select _Time & language_ -> _Language & region_.
- Manipulate (reorder) the _Preferred languages_ list.
- Start DOSBox - it should use the first language from the list we have a translation for.

6. General regression testing (to test the refactoring).
Set the following options in the config file:
- `locale_period = modern`
- `country = auto`
- `keyboard_layout = auto`
Check that detected locale settings (check the `LOCALE` logs) are the same as before (translation can differ - see tests above).


The change has been manually tested on:



# Checklist

I have:
